proof_of_concept/registry.py

```diff
@@ -5,194 +5,279 @@
-            namespace: Namespace to look up
+class AssetStoreDescription(RegisteredObject):
+    """Describes an asset store to the rest of the DDM.
 
-        Return:
-            Name of the owner of the namespace.
+    Attributes:
+        site: The site at which this store is located.
+        store: The store service.
+
+    """
+    def __init__(
+            self, site: SiteDescription, store: IAssetStore
+            ) -> None:
+        """Create an AssetStoreDescription.
+
+        Args:
+            site: The site at which this store is located.
+            store: The store service.
 
-        Raises:
-            KeyError: If no namespace with the given name is
-                    registered.
         """
-        return self._ns_owners[namespace]
+        self.site = site
+        self.store = store
 
-    def get_public_key(self, party: str) -> RSAPublicKey:
-        """Returns the public key of the given party.
 
-        Args:
-            party: Name of the party to look up.
+class PolicyServerDescription(RegisteredObject):
+    """Describes a policy server to the rest of the DDM.
 
-        Return:
-            The public key of the given party.
+    Attributes:
+        namespace: The namespace governed by this server.
+        site: The site at which this server is located.
 
-        Raises:
-            KeyError: If no party with the given name is registered.
-        """
-        return self._party_keys[party]
+    """
+    def __init__(
+            self, site: SiteDescription, namespace: NamespaceDescription,
+            server: IPolicyServer) -> None:
+        """Create a PolicyServerDescription.
 
-    def list_runners(self) -> List[str]:
-        """List names of all registered runners.
+        Args:
+            site: The site at which this server is located.
+            namespace: The namespace governed by this server.
+            server: The policy server.
 
-        Returns:
-            A list of names as strings.
         """
-        return list(self._runners.keys())
+        self.site = site
+        self.namespace = namespace
+        self.server = server
 
-    def get_runner(self, name: str) -> ILocalWorkflowRunner:
-        """Look up a LocalWorkflowRunner.
 
-        Args:
-            name: The name of the runner to look up.
+class Registry:
+    """Global registry of remote-accessible things.
 
-        Return:
-            The runner with that name.
+    Registers runners, stores, and assets. In a real system, runners
+    and stores would be identified by a URL, and use the DNS to
+    resolve. For now the registry helps with this.
+    """
+    def __init__(self) -> None:
+        """Create a new registry."""
+        self._assets = dict()           # type: Dict[str, str]
 
-        Raises:
-            KeyError: If no runner with the given name is
-                    registered.
-        """
-        return self._runners[name]
+        self._archive = ReplicableArchive[RegisteredObject]()
```

Comment:
The archive is only needed to setup the store and replication_server, so maybe not assign it to `self`?

Reply:
Good point, will change.

Reply:
Done.